src/core/device.c

```diff
@@ -1109,45 +1135,38 @@ static int device_dispatch_io(sd_device_monitor *monitor, sd_device *dev, void *
          * change events */
         ready = device_is_ready(dev);
 
+        (void) device_setup_units(m, dev, &ready_units, &not_ready_units);
+
         if (action == SD_DEVICE_REMOVE) {
                 r = swap_process_device_remove(m, dev);
                 if (r < 0)
                         log_device_warning_errno(dev, r, "Failed to process swap device remove event, ignoring: %m");
         } else if (ready) {
-                device_process_new(m, dev, sysfs);
-
                 r = swap_process_device_new(m, dev);
                 if (r < 0)
                         log_device_warning_errno(dev, r, "Failed to process swap device new event, ignoring: %m");
         }
 
-        /* Add/update additional units for all symlinks. */
-        (void) device_setup_devlink_units(m, dev, &ready_devlinks);
-
-        if (!IN_SET(action, SD_DEVICE_ADD, SD_DEVICE_REMOVE, SD_DEVICE_MOVE)) {
-                device_propagate_reload_by_sysfs(m, sysfs);
+        if (!IN_SET(action, SD_DEVICE_ADD, SD_DEVICE_REMOVE, SD_DEVICE_MOVE))
+                SET_FOREACH(d, ready_units)
+                        device_propagate_reload(m, d);
 
-                STRV_FOREACH(devlink, ready_devlinks)
-                        device_propagate_reload_by_name(m, *devlink);
-        }
-
-        if (ready || !strv_isempty(ready_devlinks))
+        if (!set_isempty(ready_units))
                 manager_dispatch_load_queue(m);
 
         if (action == SD_DEVICE_REMOVE)
                 /* If we get notified that a device was removed by udev, then it's completely gone, hence
                  * unset all found bits */
                 device_update_found_by_sysfs(m, sysfs, DEVICE_NOT_FOUND, DEVICE_FOUND_MASK);
```

Comment:
This only handles alias units right? Where's the code that does this for devlink units?

Reply:
No, this works for all (main, devnode, devlink, and alias) .device units still points to the removed device.